src/compositemark/errorbar.ts

```diff
@@ -1,42 +1,215 @@
+export type ErrorBarPartsMixins = {
+  [part in ErrorBarPart]?: boolean | MarkConfig
+};
+
+export interface ErrorBarConfig extends ErrorBarPartsMixins {
+  /**
+   * The center of the errorbar. Available options include:
+   * - `"mean": the mean of the data points.
+   * - `"median": the median of the data points.
+   *
+   * __Default value:__ `"mean"`.
+   */
+  center?: ErrorBarCenter;
+
+  /**
+   * The extent of the rule. Available options include:
+   * - `"ci": Extend the rule to the confidence interval of the mean.
+   * - `"stderr": The size of rule are set to the value of standard error, extending from the center.
+   * - `"stdev": The size of rule are set to the value of standard deviation, extending from the center.
+   * - `"iqr": Extend the rule to the q1 and q3.
+   *
+   * __Default value:__ `"stderr"`.
+   */
+  extent?: ErrorBarExtent;
+}
+
+export type ErrorBarDef = GenericCompositeMarkDef<ErrorBar> & ErrorBarConfig & {
+  /**
+   * Orientation of the error bar.  This is normally automatically determined, but can be specified when the orientation is ambiguous and cannot be automatically determined.
+   */
+  orient?: Orient;
+};
+
+export interface ErrorBarConfigMixins {
+  /**
+   * ErrorBar Config
+   */
+  errorbar?: ErrorBarConfig;
+}
+
+export function normalizeErrorBar(spec: GenericUnitSpec<Encoding<string>, ErrorBar | ErrorBarDef>, config: Config): NormalizedLayerSpec {
+  spec = filterUnsupportedChannels(spec, ERRORBAR);
+
   // TODO: use selection
-  const {mark: _m, selection: _sel, projection: _p, encoding, ...outerSpec} = spec;
-  const {size: _s, ...encodingWithoutSize} = encoding;
-  const {x2: _x2, y2: _y2, ...encodingWithoutX2Y2} = encoding;
-  const {x: _x, y: _y, ...encodingWithoutX_X2_Y_Y2} = encodingWithoutX2Y2;
+  const {mark, encoding, selection, projection: _p, ...outerSpec} = spec;
+  const markDef: ErrorBarDef = isMarkDef(mark) ? mark : {type: mark};
+
+  // TODO(https://github.com/vega/vega-lite/issues/3702): add selection support
+  if (selection) {
+    log.warn(log.message.selectionNotSupported('errorbar'));
+  }
+
+  const center: ErrorBarCenter = markDef.center || config.errorbar.center;
+  const extent: ErrorBarExtent = markDef.extent || ((center === 'mean') ? 'stderr' : 'iqr');
+
+  if ((center === 'median') !== (extent === 'iqr')) {
+    log.warn(`${center} is not usually used with ${extent} for error bar.`);
+  }
+
+  const orient: Orient = compositeMarkOrient(spec, ERRORBAR);
+  const {transform, continuousAxisChannelDef, continuousAxis, encodingWithoutContinuousAxis} = errorBarParams(spec, orient, center, extent);
+
+  const {size: _s, ...encodingWithoutSizeAndContinuousAxis} = encodingWithoutContinuousAxis;
 
-  if (!encoding.x2 && !encoding.y2) {
-    throw new Error('Neither x2 or y2 provided');
+  const {scale, axis} = continuousAxisChannelDef;
+
+  function errorBarPartSpec(partName: ErrorBarPart, startingPosition: string = center, endingPosition: string = undefined) {
```

Comment:
This function is still not extracted as commented in https://github.com/vega/vega-lite/pull/3588#discussion_r186473899